arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt

```diff
@@ -37,54 +36,50 @@ public fun <K, A, B> Map<K, A>.zip(other: Map<K, B>): Map<K, Pair<A, B>> =
  * and combining the elements with the given function.
  *
  * ```kotlin
- * import arrow.core.*
+ * import arrow.core.zip
+ * import io.kotest.matchers.shouldBe
  *
- * fun main(args: Array<String>) {
- *   //sampleStart
- *   val result =
- *    mapOf(1 to "A", 2 to "B").zip(mapOf(1 to "1", 2 to "2", 3 to "3")) {
- *      key, a, b -> "$key -> $a # $b"
- *    }
- *   //sampleEnd
- *   println(result)
+ * fun test() {
+ *   mapOf(1 to "A", 2 to "B").zip(mapOf(1 to "1", 2 to "2", 3 to "3")) {
+ *     key, a, b -> "$a ~ $b"
+ *   } shouldBe mapOf(1 to "A ~ 1", 2 to "B ~ 2")
  * }
  * ```
  * <!--- KNIT example-map-02.kt -->
+ * <!--- lines.isEmpty() -->
  */
-public inline fun <Key, A, B, C> Map<Key, A>.zip(other: Map<Key, B>, map: (Key, A, B) -> C): Map<Key, C> {
-  val destination = LinkedHashMap<Key, C>(size)
-  for ((key, bb) in this) {
-    Nullable.zip(other[key]) { cc -> map(key, bb, cc) }
-      ?.let { l -> destination.put(key, l) }
+public inline fun <Key, A, B, C> Map<Key, A>.zip(other: Map<Key, B>, map: (Key, A, B) -> C): Map<Key, C> =
+  buildMap(size) {
+    this@zip.forEach { (key, bb) ->
+      nullable {
```

Comment:
@serras @franciscodr actually this all suffers from the _nested nullable problem_ 😓 The previous implementation was also not correct... I will fix this implementation first, or should we move the fix to 2.0.0 since it's a breaking change in the current semantics?

Or should we fix the bug for nested null in the current version?

Reply:
I think that counts as a bug, and thus should be fixed in this version.

Reply:
Agreed